File: backend/routes/medicines.js
const express = require('express');
const router = express.Router();
const Medicine = require('../models/Medicine');
const { callGeminiAgent } = require('../services/aiService');

// Browse all medicines by letter — sorted A-Z, paginated
// GET /api/medicines/browse?letter=A&page=1&limit=24
router.get('/browse', async (req, res) => {
  try {
    const letter = req.query.letter || 'A';
    const page   = parseInt(req.query.page)  || 1;
    const limit  = parseInt(req.query.limit) || 24;
    const { total, medicines } = await Medicine.browse(letter, page, limit);
    res.json({ success: true, total, count: medicines.length, medicines });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Search medicines with pagination
// GET /api/medicines/search?q=paracetamol&page=1&limit=24
router.get('/search', async (req, res) => {
  try {
    const { q, page = 1, limit = 24 } = req.query;
    if (!q) return res.json({ success: true, total: 0, medicines: [] });

    const { total, medicines } = await Medicine.searchPaginated(q, parseInt(page), parseInt(limit));

    if (total === 0) {
      const aiResult = await callGeminiAgent('medicineAdvisor', q);
      return res.json({ success: true, total: 0, medicines: [], aiResponse: aiResult.data });
    }

    res.json({ success: true, total, count: medicines.length, medicines });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get medicine categories with counts
router.get('/categories', async (req, res) => {
  try {
    const categories = await Medicine.getCategories();
    res.json({ success: true, categories });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get medicines by category
router.get('/category/:category', async (req, res) => {
  try {
    const medicines = await Medicine.findByCategory(req.params.category);
    res.json({ success: true, count: medicines.length, medicines });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Find cheaper alternatives for a medicine
router.get('/alternatives/:name', async (req, res) => {
  try {
    const alternatives = await Medicine.findAlternatives(req.params.name);
    res.json({
      success: true,
      original: req.params.name,
      count: alternatives.length,
      alternatives,
    });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get medicine info via AI
router.post('/ask', async (req, res) => {
  try {
    const { question } = req.body;
    const result = await callGeminiAgent('medicineAdvisor', question);
    res.json({ success: true, response: result.data });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

// Get medicine by name
router.get('/:name', async (req, res) => {
  try {
    const medicine = await Medicine.findByName(req.params.name);
    if (!medicine) {
      const aiResult = await callGeminiAgent('medicineAdvisor', `Tell me about ${req.params.name}`);
      return res.json({ success: true, medicine: null, aiResponse: aiResult.data });
    }
    res.json({ success: true, medicine });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
});

module.exports = router;

File: backend/routes/medicines.py
"""
routes/medicines.py — Medicine search, letter browse, categories, alternatives, and AI medicine advisor
Mirrors backend/routes/medicines.js
"""
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from typing import Optional

from models.medicine import Medicine
from services.ai_service import call_groq_agent

router = APIRouter(prefix="/api/medicines", tags=["Medicines"])


class AskQuestionRequest(BaseModel):
    question: str


@router.get("/browse")
async def browse_medicines(
    letter: str = "A",
    page: int = 1,
    limit: int = 24,
):
    try:
        data = Medicine.browse(letter=letter, page=page, limit=limit)
        return {
            "success": True,
            "total": data["total"],
            "count": len(data["medicines"]),
            "medicines": data["medicines"],
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/search")
async def search_medicines(
    q: Optional[str] = None,
    page: int = 1,
    limit: int = 24,
):
    try:
        if not q:
            return {"success": True, "total": 0, "medicines": []}

        data = Medicine.search_paginated(q, page=page, limit=limit)
        if data["total"] == 0:
            ai_result = await call_groq_agent("medicineAdvisor", q)
            return {"success": True, "total": 0, "medicines": [], "aiResponse": ai_result.get("data", "")}

        return {
            "success": True,
            "total": data["total"],
            "count": len(data["medicines"]),
            "medicines": data["medicines"],
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/categories")
async def get_categories():
    try:
        categories = Medicine.get_categories()
        return {"success": True, "categories": categories}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/category/{category}")
async def get_medicines_by_category(category: str):
    try:
        medicines = Medicine.find_by_category(category)
        return {"success": True, "count": len(medicines), "medicines": medicines}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/alternatives/{name}")
async def get_alternatives(name: str):
    try:
        alternatives = Medicine.find_alternatives(name)
        return {
            "success": True,
            "original": name,
            "count": len(alternatives),
            "alternatives": alternatives,
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/ask")
async def ask_medicine_ai(payload: AskQuestionRequest):
    try:
        result = await call_groq_agent("medicineAdvisor", payload.question)
        return {"success": True, "response": result.get("data", "")}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{name}")
async def get_medicine_by_name(name: str):
    try:
        med = Medicine.find_by_name(name)
        if not med:
            ai_result = await call_groq_agent("medicineAdvisor", f"Tell me about {name}")
            return {"success": True, "medicine": None, "aiResponse": ai_result.get("data", "")}
        return {"success": True, "medicine": med}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
